File: 953 - Verifying an Alien Dictionary.java
class Solution {
    public boolean isAlienSorted(String[] words, String order) {
        int[] idx = new int[26];
        for (int i = 0; i < 26; i++) {
            idx[order.charAt(i) - 'a'] = i;
        }
        for (int i = 0; i < words.length - 1; i++) {
            String word1 = words[i], word2 = words[i + 1];
            boolean less = false;
            if (!word1.equals(word2)) {
                int n = Math.min(word1.length(), word2.length());
                for (int j = 0; j < n; j++) {
                    if (word1.charAt(j) != word2.charAt(j)) {
                        if(idx[word1.charAt(j) - 'a'] > idx[word2.charAt(j) - 'a']) {
                            return false;
                        } else {
                            less = true;
                            break;
                        }
                    }
                }
                if (word1.length() > word2.length() && !less) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: 953 - Verifying an Alien Dictionary.js
/**
 * @param {string[]} words
 * @param {string} order
 * @return {boolean}
 */
var isAlienSorted = function(words, order) {
    let idx = new Array(26);
    for (let i = 0; i < 26; i++) {
        idx[order[i].charCodeAt() - 'a'.charCodeAt()] = i;
    }
    for (let i = 0; i < words.length - 1; i++) {
        let word1 = words[i], word2 = words[i + 1];
        if (word1 != word2) {
            let less = false;
            let n = Math.min(word1.length, word2.length);
            for (let j = 0; j < n; j++) {
                if (word1[j] != word2[j]) {
                    if (idx[word1[j].charCodeAt() - 'a'.charCodeAt()] > idx[word2[j].charCodeAt() - 'a'.charCodeAt()]) {
                        return false;
                    } else {
                        less = true;
                        break;
                    }
                }
            }
            if (word1.length > word2.length && !less) {
                return false;
            }
        }
    }
    return true;
};

let a = ["word","world","row"], b = "worldabcefghijkmnpqstuvxyz";
isAlienSorted(a, b);